diff --git a/django_cas_ng/__init__.py b/django_cas_ng/__init__.py
index df71356..a41f80a 100644
--- a/django_cas_ng/__init__.py
+++ b/django_cas_ng/__init__.py
@@ -8,6 +8,7 @@ from django.utils.translation import ugettext_lazy as _
 __all__ = []
 
 _DEFAULTS = {
+    'CAS_ACCEPT_PROXY_URL_FROM_HEADER': None,
     'CAS_ADMIN_PREFIX': None,
     'CAS_CREATE_USER': True,
     'CAS_EXTRA_LOGIN_PARAMS': None,
diff --git a/django_cas_ng/utils.py b/django_cas_ng/utils.py
index 1609289..5f295f3 100644
--- a/django_cas_ng/utils.py
+++ b/django_cas_ng/utils.py
@@ -44,6 +44,10 @@ def get_service_url(request, redirect_to=None):
     """Generates application django service URL for CAS"""
     if hasattr(django_settings, 'CAS_ROOT_PROXIED_AS'):
         service = django_settings.CAS_ROOT_PROXIED_AS + request.path
+    elif (hasattr(django_settings, 'CAS_ACCEPT_PROXY_URL_FROM_HEADER') and
+                django_settings.CAS_ACCEPT_PROXY_URL_FROM_HEADER and
+                request.META.get('HTTP_'+django_settings.CAS_ACCEPT_PROXY_URL_FROM_HEADER, False)):
+        service = request.META.get('HTTP_'+django_settings.CAS_ACCEPT_PROXY_URL_FROM_HEADER)
     else:
         protocol = get_protocol(request)
         host = request.get_host()
diff --git a/django_cas_ng/views.py b/django_cas_ng/views.py
index 3201936..aeeacc3 100644
--- a/django_cas_ng/views.py
+++ b/django_cas_ng/views.py
@@ -47,7 +47,11 @@ class LoginView(View):
         :param next_page:
         :return:
         """
-        return HttpResponseRedirect(next_page)
+        # return HttpResponseRedirect(next_page)
+        from varda.custom_login_views import set_cookie
+        response = HttpResponseRedirect(next_page)
+        set_cookie(request, response, next_page)
+        return response
 
     def post(self, request):
         if request.POST.get('logoutRequest'):
